### ISSUE TITLE
ci: init

### ISSUE THREAD

**Issue by oSumAtrIX** [revanced/revanced-manager-compose#49]
## About

Currently, this repository lacks CI/CD.

## Solution

Add CI/CD to run tests, build and deploy releases.

## Additional context

- Prereleases should be dispatched manually for development branches.
- Releases should be dispatched automatically when commits are pushed to main.
- CalVer should be used for versioning
- A changelog should be generated for releases compiled from the commits following the semantic commits specification

**Comment by PalmDevs** [comment]
Decided on this CalVer version format: `YYYY.0W.MICRO` (eg. 2023.30.1 (`Year 2023, Week 30, Revision (patch) 1`)

**Comment by oSumAtrIX** [comment]
Why use week and lose the semantics of major?

**Comment by oSumAtrIX** [comment]
`YYYY.MINOR.PATCH.MICRO` keeps the semantics that ReVanced Manager currently also has, which are minor and patch. It gets rid of major which is the only problem that we had with SemVer. How is weeks instead of patch better?

**Comment by Axelen123** [comment]
Why do we need `MICRO` or `MINOR`? what purpose do they serve in the context of a user application?

**Comment by oSumAtrIX** [comment]
The purpose is the same as that of SemVer. `PATCH` delivers fixes for example, whereas `MICRO` may be changed intraday. Though on second thought, `MICRO` can be even left out, since all the commit types that we would use always either bump `MINOR` or `PATCH` therefore `MICRO` can be omitted

**Comment by Axelen123** [comment]
If thats the scheme we are going for then yes, `MICRO` can and should be omitted.

Will our users know how to interpret semver, `MINOR` or `PATCH`?

**Comment by oSumAtrIX** [comment]
Yes, more than `YYYY.0W.MICRO` because it is not easy to understand 0W because it changes weekly so you have to observe it over weeks to understand it.

Wheres as you can directly see a pattern in minor and patch bumps

**Comment by Axelen123** [comment]
Months could be used instead of weeks then. That would make it look more like a date, which everyone is familiar with.

**Comment by oSumAtrIX** [comment]
Month or week, we occupy the minor version. Our current issue with SemVer is major. How do we solve it? Not by getting rid of minor.

**Comment by oSumAtrIX** [comment]
@PalmDevs I have answered this later on here: https://github.com/revanced/revanced-manager-compose/issues/49#issuecomment-1630350259

**Comment by alexandreteles** [comment]
I'm putting this in the backlog for better tracking.

Despite being unable to decide which versioning scheme to use, the lack of CI makes it hard to follow the development and provide feedback for users not currently involved in directly developing the product, which is detrimental to the product's lifecycle.

You can simply set up the CI and upload the relevant build files as build artifacts instead of releasing them until you have figured this out.

**Comment by Axelen123** [comment]
@PalmDevs @alexandreteles @validcube any objections to the current format proposed by @oSumAtrIX?
I don't have any

**Comment by PalmDevs** [comment]
> @PalmDevs @alexandreteles @validcube any objections to the current format proposed by @oSumAtrIX?
> I don't have any

What's the decided format again? The whole discussion didn't make it too clear.

**Comment by Axelen123** [comment]
> What's the decided format again? The whole discussion didn't make it too clear.

YYYY.MINOR.PATCH

**Comment by alexandreteles** [comment]
Only that I think that pre-release versions should be built on push for dev so they are always on sync with the latest development efforts.

**Comment by Axelen123** [comment]
> Only that I think that pre-release versions should be built on push for dev so they are always on sync with the latest development efforts.

Isn't [the workflow we already have](https://github.com/ReVanced/revanced-manager-compose/blob/dev/.github/workflows/release.yml) good enough for that?

**Comment by oSumAtrIX** [comment]
I think what @alexandreteles refers to is that CI triggers automatically on any push on dev but also on main. Currently it only triggers on dev: https://github.com/ReVanced/revanced-manager-compose/blob/dev/.github/workflows/release.yml#L3-L9

**Comment by alexandreteles** [comment]
> YYYY.MINOR.PATCH

@Axelen123 Did you mean `YYYY.MINOR.MICRO`? Yeah, this is fine. Pre-release versions should include the `-dev` modifier; everything will be fine, clear, and understandable.

- 22.30.4
- 23.30.1
- 23.30.2-dev
- 23.30.3
- 23.31.0

I prefer `YY` for its shortness, but both patterns are good because they are clear:

1. Whatever is released this year is newer than what was released last year.
2. Whatever is released in the same year but has a bigger `minor` is newer.
3. Whatever is released in the same year, with the same `minor` but a higher `micro`, is newer.
4. The `-dev` modifier is a text tag differentiating between production-ready and development builds.
